File: Services/ConfigurationService.cs
using System.Text.Json;
using OpenAI.Chat;
using Azure.AI.OpenAI;
using System.ClientModel;
using WriteCommit.Constants;
using WriteCommit.Models;

namespace WriteCommit.Services;

public class ConfigurationService
{
    private readonly string _configDirectory;
    private readonly string _configFilePath;

    public ConfigurationService()
    {
        var homeDirectory = Environment.GetFolderPath(Environment.SpecialFolder.UserProfile);
        _configDirectory = Path.Combine(homeDirectory, ".writecommit");
        _configFilePath = Path.Combine(_configDirectory, "config.json");
    }

    /// <summary>
    /// Loads the configuration from file
    /// </summary>
    public async Task<AppConfiguration?> LoadConfigurationAsync()
    {
        if (!File.Exists(_configFilePath))
        {
            return null;
        }

        try
        {
            var json = await File.ReadAllTextAsync(_configFilePath);
            return JsonSerializer.Deserialize<AppConfiguration>(json);
        }
        catch
        {
            // If there's any error reading or parsing the config, return null
            return null;
        }
    }

    /// <summary>
    /// Saves the configuration to file
    /// </summary>
    public async Task SaveConfigurationAsync(AppConfiguration config)
    {
        // Ensure directory exists
        Directory.CreateDirectory(_configDirectory);

        var options = new JsonSerializerOptions { WriteIndented = true };

        var json = JsonSerializer.Serialize(config, options);
        await File.WriteAllTextAsync(_configFilePath, json);

        // Set appropriate file permissions on Unix systems
        if (OperatingSystem.IsLinux() || OperatingSystem.IsMacOS())
        {
            SetUnixFilePermissions(_configFilePath);
        }
    }

    /// <summary>
    /// Gets the OpenAI API key from environment variable or config file
    /// </summary>
    public async Task<string?> GetOpenAiApiKeyAsync()
    {
        // First check environment variable (highest priority)
        var envApiKey = Environment.GetEnvironmentVariable("OPENAI_API_KEY");
        if (!string.IsNullOrEmpty(envApiKey))
        {
            return envApiKey;
        }

        // Then check config file
        var config = await LoadConfigurationAsync();
        return config?.OpenAiApiKey;
    }

    /// <summary>
    /// Gets the configured OpenAI endpoint or null if not set
    /// </summary>
    public async Task<string?> GetOpenAiEndpointAsync()
    {
        var config = await LoadConfigurationAsync();
        return config?.OpenAiEndpoint;
    }

    /// <summary>
    /// Gets the configured default model or null if not set
    /// </summary>
    public async Task<string?> GetDefaultModelAsync()
    {
        var config = await LoadConfigurationAsync();
        return config?.DefaultModel;
    }

    /// <summary>
    /// Returns true if configuration specifies Azure OpenAI usage
    /// </summary>
    public async Task<bool> UseAzureOpenAIAsync()
    {
        var config = await LoadConfigurationAsync();
        return config?.UseAzureOpenAI ?? false;
    }

    /// <summary>
    /// Prompts user to enter and save their OpenAI API key
    /// </summary>
    public async Task<bool> SetupApiKeyAsync(bool verbose = false)
    {
        Console.WriteLine("WriteCommit Setup");
        Console.WriteLine("=================");
        Console.WriteLine();
        Console.WriteLine("Please enter your OpenAI API key (or Azure OpenAI key).");
        Console.WriteLine("You can get one from: https://platform.openai.com/api-keys or your Azure portal");
        Console.WriteLine();
        Console.Write("API Key (leave blank if not required): ");

        // Read API key with masked input
        var apiKey = ReadMaskedInput();
        Console.WriteLine();

        if (string.IsNullOrWhiteSpace(apiKey))
        {
            apiKey = null;
        }

        // Ask if using Azure OpenAI
        Console.Write("Use Azure OpenAI service? (y/N): ");
        var azureInput = Console.ReadLine()?.Trim().ToLowerInvariant();
        bool useAzure = azureInput == "y" || azureInput == "yes";

        // Prompt for endpoint and model/deployment
        Console.Write(
            useAzure
                ? "Azure endpoint (e.g. https://your-resource.openai.azure.com): "
                : "Endpoint (default: https://api.openai.com/v1): "
        );
        var endpointInput = Console.ReadLine()?.Trim();
        var endpoint = string.IsNullOrWhiteSpace(endpointInput)
            ? (useAzure ? "https://your-resource.openai.azure.com" : "https://api.openai.com/v1")
            : endpointInput;

        Console.Write(
            useAzure
                ? "Deployment name (default: gpt-4o-mini): "
                : "Default model (default: gpt-4o-mini): "
        );
        var modelInput = Console.ReadLine()?.Trim();
        var model = string.IsNullOrWhiteSpace(modelInput) ? "gpt-4o-mini" : modelInput;

        // Load existing config or create new one
        var config = await LoadConfigurationAsync() ?? new AppConfiguration();
        config.OpenAiApiKey = apiKey;
        config.OpenAiEndpoint = endpoint;
        config.DefaultModel = model;
        config.UseAzureOpenAI = useAzure;

        // Save configuration
        await SaveConfigurationAsync(config);

        Console.WriteLine($"✅ Configuration saved to {_configFilePath}");

        if (verbose)
        {
            Console.WriteLine("Configuration saved successfully.");
            Console.WriteLine($"Config location: {_configFilePath}");
        }

        // Optionally test the API key
        Console.WriteLine();
        Console.Write("Would you like to test the API key? (y/N): ");
        var testResponse = Console.ReadLine()?.Trim().ToLowerInvariant();

        if ((testResponse == "y" || testResponse == "yes") && !string.IsNullOrEmpty(apiKey))
        {
            return await TestApiKeyAsync(apiKey, useAzure, endpoint, model, verbose);
        }

        return true;
    }

    /// <summary>
    /// Tests if the API key is valid by making a simple request
    /// </summary>
    private async Task<bool> TestApiKeyAsync(string? apiKey, bool useAzure, string endpoint, string model, bool verbose)
    {
        Console.WriteLine("Testing API key...");

        try
        {
            ChatClient testClient;
            if (useAzure)
            {
                var azureClient = new Azure.AI.OpenAI.AzureOpenAIClient(new Uri(endpoint), new ApiKeyCredential(apiKey));
                testClient = azureClient.GetChatClient(model);
            }
            else
            {
                var options = new OpenAI.OpenAIClientOptions { Endpoint = new Uri(endpoint) };
                testClient = new ChatClient(model, new ApiKeyCredential(apiKey), options);
            }
            var messages = new List<ChatMessage>
            {
                new SystemChatMessage("You are a helpful assistant."),
                new UserChatMessage("Say 'API key is valid' if you can read this."),
            };

            var response = await testClient.CompleteChatAsync(messages);

            if (response.Value != null)
            {
                Console.WriteLine("✅ API key is valid and working!");
                return true;
            }
            else
            {
                Console.WriteLine("❌ API key test failed: No response received");
                return false;
            }
        }
        catch (Exception ex)
        {
            Console.WriteLine($"❌ API key test failed: {ex.Message}");
            if (verbose)
            {
                Console.WriteLine($"Full error: {ex}");
            }
            return false;
        }
    }

    /// <summary>
    /// Sets Unix file permissions to 600 (owner read/write only)
    /// </summary>
    private void SetUnixFilePermissions(string filePath)
    {
        // Only try to set Unix file permissions on Unix-based systems
        if (OperatingSystem.IsLinux() || OperatingSystem.IsMacOS())
        {
            try
            {
#if !WINDOWS
                File.SetUnixFileMode(filePath, UnixFileMode.UserRead | UnixFileMode.UserWrite);
#endif
            }
            catch
            {
                // Silently ignore if setting permissions fails
            }
        }
    }

    /// <summary>
    /// Reads input from console with masking
    /// </summary>
    private string ReadMaskedInput()
    {
        var input = "";
        ConsoleKeyInfo key;

        do
        {
            key = Console.ReadKey(true);

            if (key.Key != ConsoleKey.Backspace && key.Key != ConsoleKey.Enter)
            {
                input += key.KeyChar;
                Console.Write("*");
            }
            else if (key.Key == ConsoleKey.Backspace && input.Length > 0)
            {
                input = input.Substring(0, input.Length - 1);
                Console.Write("\b \b");
            }
        } while (key.Key != ConsoleKey.Enter);

        return input;
    }
}

File: src/services/ConfigurationService.ts
import { promises as fs } from 'fs';
import { join } from 'path';
import { homedir } from 'os';
import { AppConfiguration } from '../models';
import OpenAI from 'openai';
import * as readline from 'readline';

export class ConfigurationService {
  private readonly configDirectory: string;
  private readonly configFilePath: string;

  constructor() {
    const homeDirectory = homedir();
    this.configDirectory = join(homeDirectory, '.writecommit');
    this.configFilePath = join(this.configDirectory, 'config.json');
  }

  /**
   * Loads the configuration from file
   */
  async loadConfiguration(): Promise<AppConfiguration | null> {
    try {
      await fs.access(this.configFilePath);
      const json = await fs.readFile(this.configFilePath, 'utf8');
      return JSON.parse(json) as AppConfiguration;
    } catch {
      // If there's any error reading or parsing the config, return null
      return null;
    }
  }

  /**
   * Saves the configuration to file
   */
  async saveConfiguration(config: AppConfiguration): Promise<void> {
    // Ensure directory exists
    await fs.mkdir(this.configDirectory, { recursive: true });

    const json = JSON.stringify(config, null, 2);
    await fs.writeFile(this.configFilePath, json, 'utf8');

    // Set appropriate file permissions on Unix systems
    if (process.platform === 'linux' || process.platform === 'darwin') {
      await this.setUnixFilePermissions(this.configFilePath);
    }
  }

  /**
   * Gets the OpenAI API key from environment variable or config file
   */
  async getOpenAiApiKey(): Promise<string | null> {
    // First check environment variable (highest priority)
    const envApiKey = process.env.OPENAI_API_KEY;
    if (envApiKey) {
      return envApiKey;
    }

    // Then check config file
    const config = await this.loadConfiguration();
    return config?.openai_api_key ?? null;
  }

  /**
   * Gets the configured OpenAI endpoint or null if not set
   */
  async getOpenAiEndpoint(): Promise<string | null> {
    const config = await this.loadConfiguration();
    return config?.openai_endpoint ?? null;
  }

  /**
   * Gets the configured default model or null if not set
   */
  async getDefaultModel(): Promise<string | null> {
    const config = await this.loadConfiguration();
    return config?.default_model ?? null;
  }

  /**
   * Returns true if configuration specifies Azure OpenAI usage
   */
  async useAzureOpenAI(): Promise<boolean> {
    const config = await this.loadConfiguration();
    return config?.use_azure_openai ?? false;
  }

  /**
   * Prompts user to enter and save their OpenAI API key
   */
  async setupApiKey(verbose = false): Promise<boolean> {
    const rl = readline.createInterface({
      input: process.stdin,
      output: process.stdout,
    });

    try {
      console.log('WriteCommit Setup');
      console.log('=================');
      console.log();
      console.log('Please enter your OpenAI API key (or Azure OpenAI key).');
      console.log('You can get one from: https://platform.openai.com/api-keys or your Azure portal');
      console.log();

      const apiKey = await this.readMaskedInput('API Key (leave blank if not required): ');
      console.log();

      const azureInput = await this.question(rl, 'Use Azure OpenAI service? (y/N): ');
      const useAzure = azureInput.trim().toLowerCase() === 'y' || azureInput.trim().toLowerCase() === 'yes';

      const endpointPrompt = useAzure
        ? 'Azure endpoint (leave blank to use OpenAI): '
        : 'Endpoint (default: https://api.openai.com/v1): ';
      
      const endpointInput = await this.question(rl, endpointPrompt);
      let endpoint: string;
      let finalUseAzure = useAzure;

      if (!endpointInput.trim()) {
        endpoint = 'https://api.openai.com/v1';
        // If no Azure endpoint provided, fall back to OpenAI
        if (useAzure) {
          finalUseAzure = false;
        }
      } else {
        endpoint = endpointInput.trim();
      }

      const modelPrompt = finalUseAzure
        ? 'Deployment name (default: gpt-4o-mini): '
        : 'Default model (default: gpt-4o-mini): ';
      
      const modelInput = await this.question(rl, modelPrompt);
      const model = modelInput.trim() || 'gpt-4o-mini';

      // Load existing config or create new one
      const config = (await this.loadConfiguration()) ?? {};
      config.openai_api_key = apiKey || undefined;
      config.openai_endpoint = endpoint;
      config.default_model = model;
      config.use_azure_openai = finalUseAzure;

      // Save configuration
      await this.saveConfiguration(config);

      console.log(`✅ Configuration saved to ${this.configFilePath}`);

      if (verbose) {
        console.log('Configuration saved successfully.');
        console.log(`Config location: ${this.configFilePath}`);
      }

      // Optionally test the API key
      console.log();
      const testResponse = await this.question(rl, 'Would you like to test the API key? (y/N): ');

      if ((testResponse.trim().toLowerCase() === 'y' || testResponse.trim().toLowerCase() === 'yes') && apiKey) {
        return await this.testApiKey(apiKey, finalUseAzure, endpoint, model, verbose);
      }

      return true;
    } finally {
      rl.close();
    }
  }

  /**
   * Tests if the API key is valid by making a simple request
   */
  private async testApiKey(
    apiKey: string,
    useAzure: boolean,
    endpoint: string,
    model: string,
    verbose: boolean
  ): Promise<boolean> {
    console.log('Testing API key...');

    try {
      const openai = new OpenAI({
        apiKey: apiKey,
        baseURL: endpoint.endsWith('/v1') ? endpoint : `${endpoint}/v1`,
      });

      const response = await openai.chat.completions.create({
        model: model,
        messages: [
          { role: 'system', content: 'You are a helpful assistant.' },
          { role: 'user', content: 'Say "API key is valid" if you can read this.' },
        ],
        max_tokens: 50,
      });

      if (response.choices && response.choices.length > 0) {
        console.log('✅ API key is valid and working!');
        return true;
      } else {
        console.log('❌ API key test failed: No response received');
        return false;
      }
    } catch (error) {
      console.log(`❌ API key test failed: ${error instanceof Error ? error.message : 'Unknown error'}`);
      if (verbose) {
        console.log(`Full error: ${error}`);
      }
      return false;
    }
  }

  /**
   * Sets Unix file permissions to 600 (owner read/write only)
   */
  private async setUnixFilePermissions(filePath: string): Promise<void> {
    // Only try to set Unix file permissions on Unix-based systems
    if (process.platform === 'linux' || process.platform === 'darwin') {
      try {
        await fs.chmod(filePath, 0o600);
      } catch {
        // Silently ignore if setting permissions fails
      }
    }
  }

  /**
   * Reads input from console with masking
   */
  private async readMaskedInput(prompt: string): Promise<string> {
    return new Promise((resolve) => {
      process.stdout.write(prompt);
      process.stdin.setRawMode(true);
      process.stdin.resume();
      process.stdin.setEncoding('utf8');

      let input = '';

      const onData = (char: string) => {
        switch (char) {
          case '\n':
          case '\r':
          case '\u0004': // Ctrl+D
            process.stdin.setRawMode(false);
            process.stdin.pause();
            process.stdin.off('data', onData);
            console.log();
            resolve(input);
            break;
          case '\u0003': // Ctrl+C
            process.exit();
            break;
          case '\u007f': // Backspace
            if (input.length > 0) {
              input = input.slice(0, -1);
              process.stdout.write('\b \b');
            }
            break;
          default:
            input += char;
            process.stdout.write('*');
            break;
        }
      };

      process.stdin.on('data', onData);
    });
  }

  private question(rl: readline.Interface, prompt: string): Promise<string> {
    return new Promise((resolve) => {
      rl.question(prompt, resolve);
    });
  }
}